Core & Internals:  Changes core to exclusively use rse_id. API layer acts as a boundary where rse name is converted to rse_id and vice versa; Fix #2500

AndrewLister-STFC opened the pull request:
Core & Internals:  Changes core to exclusively use rse_id. API layer acts as a boundary where rse name is converted to rse_id and vice versa; Fix #2500

AndrewLister-STFC committed:
Core & Internals:  Changes core to exclusively use rse_id. API layer acts as a boundary where rse name is converted to rse_id and vice versa; Fix #2500

mlassnig commented on lib/rucio/api/account_limit.py:
```diff
@@ -115,14 +120,15 @@ def get_account_usage(account, rse, issuer):
 
-    kwargs = {'account': account, 'rse': rse}
+    rse_id = None
+    print('get_acc_usage')
```
extra print

AndrewLister-STFC replied:
Fixed in new commit

mlassnig commented on lib/rucio/api/replica.py:
```diff
@@ -118,13 +147,28 @@ def list_replicas(dids, schemes=None, unavailable=False, request_id=None,
+                                     sign_urls=sign_urls, signature_lifetime=signature_lifetime,
+                                     resolve_archives=resolve_archives, resolve_parents=resolve_parents)
+
+    for rep in replicas:
```
does this overwrite the field? the existing list_replicas response should not be modified (stuff can be added though). i'm not really following the code below

AndrewLister-STFC replied:
Yes, "list_replicas" returns a dict with stucture dependant on keys which were the rse name (replicas['rses']['MOCK'], replicas['rses']['MOCK2']...). 
I changed it to rse_id so that if any other part of core needs to rely on it its still ids. 
Its then the job in api to translate those keys.
So it takes the result and replaces these keys by creating a new entry for 'rses' and 'states'. 

Why shouldn't it be modified?

mlassnig replied:
This is the API here, there are clients that rely on the RSE being the RSE name.

AndrewLister-STFC replied:
Do those clients use core? The rse is id in core, then before it returns from api is convereted to name

mlassnig replied:
Ah, no, if the reply of the RSE stays the same that's fine. Be careful though that the lists in here are all sorted, so rewriting them must not change the order.

AndrewLister-STFC replied:
I hadn't spotted that they're sorted. This bit won't change the order but due to changes in the core the rses will be sorted by rse_id. But I thought dicts were unordered anyway?

mlassnig commented on lib/rucio/core/rule.py:
```diff
@@ -1850,7 +1850,7 @@ def update_rules_for_bad_replica(scope, name, rse_id, nowait=False, session=None
         session.query(models.RSEFileAssociation).filter(models.RSEFileAssociation.scope == scope, models.RSEFileAssociation.name == name, models.RSEFileAssociation.rse_id == rse_id).update({'state': ReplicaState.COPYING})
     else:
-        logging.info('File %s:%s at site %s has no locks. Will be deleted now.', scope, name, rse)
+        logging.info('File %s:%s at site %s has no locks. Will be deleted now.', scope, name, rse_id)
```
for all the logging it would be good to have both RSE name and RSE ID

AndrewLister-STFC replied:
Good point, I can update it to a format like `logging.info('File %s:%s at site %s(%s) has no locks. Will be deleted now.', scope, name, get_rse_name(rse_id=rse_id, session=session), rse_id) ` throughout core

bari12 replied:
We could even leave rse name for logging. I mean, in the case of multi-VO, the VO-Name would be encoded in the scope anyway, so it is somewhat clear which RSE is concerned.

AndrewLister-STFC replied:
If we're using the external representation for rse, it'd be cleaner to use the external representation for all scope and accounts too and explicitly stating the vo. In which case just the name is fine here and when VO is added, it will need to be inserted into all the logging.

AndrewLister-STFC replied:
I've updated the core logging to use rse names instead of ids.

mlassnig reviewed (commented):
the mother of all patches ;-) looks fairly good to me, but a few more pairs of eyes will certainly be useful. testing this will be interesting.

just one question, i see that you've used dicts like `rse["id"]` in many places. would it make sense to just use namedtuples so it can be `rse.id` ? a bit easier on the eyes.

AndrewLister-STFC commented:
The next patch will be a similar size for the account/scope version of this...

For rse["id"] vs rse.id I tried to just follow the lead of the code that was already there but I agree that named tuples look better.

bari12 commented:
There is one error in the syntax test `lib/rucio/api/did.py:140:1: W293 blank line contains whitespace`

AndrewLister-STFC committed:
Minor changes from pull request comments

AndrewLister-STFC committed:
Change logger calls to return rse names

AndrewLister-STFC committed:
Merge master into branch as outdated

AndrewLister-STFC committed:
Fix issues from linter

AndrewLister-STFC committed:
Exceptions use rse name rather than id

AndrewLister-STFC committed:
RSE names returned with id from core

AndrewLister-STFC committed:
Linting fixes

AndrewLister-STFC committed:
Remove excessive API conversions as now in core

AndrewLister-STFC committed:
Merge main in to keep updated

AndrewLister-STFC committed:
Small fixes for tests

AndrewLister-STFC committed:
Remove unused var

AndrewLister-STFC committed:
Core & Internals: Merge latest changes into branch; Fix #2500
Merge branch 'master' into patch-2500-Switch_to_rse_id_for_core

Conflicts:
	lib/rucio/core/exporter.py
	lib/rucio/core/importer.py
	lib/rucio/core/replica.py
	lib/rucio/core/rse.py
	lib/rucio/tests/test_import_export.py

AndrewLister-STFC committed:
Core & Internals: Fix minor issues with importer tests; Fix #2500

AndrewLister-STFC committed:
Core & Internals: Merge and fix conflicts; Fix #2500

Conflicts:
	lib/rucio/tests/test_import_export.py

AndrewLister-STFC committed:
Core & Internals: Fix linting complaints; Fix #2551

AndrewLister-STFC committed:
Core & Internals: Small changes requested after review; Fix #2500
